File: crates/durable-java/src/main/java/cleat/Saga.java
package cleat;

import java.util.ArrayList;
import java.util.List;

/**
 * A simple saga (compensating transaction) framework for cleat workflows.
 * <p>
 * A saga coordinates a sequence of steps, each with a forward action and a
 * compensating action.  If any forward step fails, all previously completed
 * steps are compensated in reverse order.
 * <p>
 * Usage:
 * <pre>{@code
 * Saga saga = new Saga()
 *     .addStep("reserve inventory",
 *         h -> h.durableCall("inv", "Reserve", input).getValue(),
 *         h -> h.durableCall("inv", "Release", input))
 *     .addStep("charge payment",
 *         h -> h.durableCall("pay", "Charge", input).getValue(),
 *         h -> h.durableCall("pay", "Refund", input));
 *
 * saga.run(hostCalls);
 * }</pre>
 */
public class Saga {

    @FunctionalInterface
    public interface SagaFunction {
        String apply(HostCalls h) throws Exception;
    }

    @FunctionalInterface
    public interface SagaCompensator {
        void accept(HostCalls h) throws Exception;
    }

    /**
     * A single saga step with a description, a forward action, and a
     * compensating action.
     */
    public static class Step {
        /** Human-readable description of this step. */
        public final String description;

        /** The forward action to execute. */
        public final SagaFunction forward;

        /** The compensating action to run on rollback. */
        public final SagaCompensator compensate;

        /**
         * Construct a new saga step.
         *
         * @param description human-readable description
         * @param forward     the forward action
         * @param compensate  the compensating action
         */
        public Step(String description, SagaFunction forward, SagaCompensator compensate) {
            this.description = description;
            this.forward = forward;
            this.compensate = compensate;
        }
    }

    private final List<Step> steps;

    /**
     * Construct a new Saga with no steps.
     * <p>
     * Steps can be added via {@link #addStep(String, SagaFunction, SagaCompensator)}
     * for chaining, or use {@link #builder()} for the builder pattern.
     */
    public Saga() {
        this.steps = new ArrayList<>();
    }

    /**
     * Construct a new Saga with the given steps.
     *
     * @param steps the steps for this saga
     */
    private Saga(List<Step> steps) {
        this.steps = new ArrayList<>(steps);
    }

    // ========================================================================
    // Builder
    // ========================================================================

    /**
     * Create a new {@link SagaBuilder} for constructing a Saga.
     * <p>
     * Usage:
     * <pre>{@code
     * Saga saga = Saga.builder()
     *     .addStep("bookFlight",
     *         h -> bookFlight(h),
     *         h -> cancelFlight(h))
     *     .addStep("chargePayment",
     *         h -> chargePayment(h),
     *         h -> refundPayment(h))
     *     .build();
     * saga.execute(h);
     * }</pre>
     *
     * @return a new {@link SagaBuilder}
     */
    public static SagaBuilder builder() {
        return new SagaBuilder();
    }

    /**
     * Builder for constructing a {@link Saga} with a fluent API.
     */
    public static class SagaBuilder {
        private final List<Step> steps = new ArrayList<>();

        private SagaBuilder() {
        }

        /**
         * Add a step to the saga.
         *
         * @param description human-readable description of the step
         * @param forward     the forward action
         * @param compensate  the compensating action
         * @return {@code this} for chaining
         */
        public SagaBuilder addStep(String description, SagaFunction forward, SagaCompensator compensate) {
            steps.add(new Step(description, forward, compensate));
            return this;
        }

        /**
         * Build the {@link Saga} with the configured steps.
         *
         * @return a new {@link Saga} instance
         */
        public Saga build() {
            return new Saga(steps);
        }
    }

    // ========================================================================
    // Direct step addition (chaining, non-builder style)
    // ========================================================================

    /**
     * Add a step to the saga.
     * <p>
     * This method supports the direct chaining style:
     * <pre>{@code
     * Saga saga = new Saga()
     *     .addStep("desc", fwd, cmp)
     *     .addStep("desc2", fwd2, cmp2);
     * }</pre>
     *
     * @param description human-readable description of the step
     * @param forward     the forward action
     * @param compensate  the compensating action
     * @return {@code this} for chaining
     */
    public Saga addStep(String description, SagaFunction forward, SagaCompensator compensate) {
        steps.add(new Step(description, forward, compensate));
        return this;
    }

    // ========================================================================
    // Execution
    // ========================================================================

    /**
     * Execute all forward steps in order.  If any step throws an exception,
     * all previously completed steps are compensated in reverse order.
     * <p>
     * Alias for {@link #run(HostCalls)}.
     *
     * @param h the HostCalls instance for this workflow execution
     * @throws Exception if a forward step fails and compensation completes
     * @throws RuntimeException if compensation itself fails (the original
     *         forward exception is added as a suppressed exception)
     */
    public void execute(HostCalls h) throws Exception {
        run(h);
    }

    // ========================================================================
    // Typed Saga (generic)
    // ========================================================================

    /**
     * A typed saga step where the forward action returns a value of type {@code T}.
     *
     * @param <T> the result type of the forward action
     */
    public static class StepTyped<T> {
        /** Human-readable description of this step. */
        public final String description;

        /** The forward action returning a value of type T. */
        public final java.util.function.Function<HostCalls, T> forward;

        /** The compensating action to run on rollback. */
        public final SagaCompensator compensate;

        /**
         * Construct a new typed saga step.
         *
         * @param description human-readable description
         * @param forward     the forward action returning T
         * @param compensate  the compensating action
         */
        public StepTyped(String description, java.util.function.Function<HostCalls, T> forward, SagaCompensator compensate) {
            this.description = description;
            this.forward = forward;
            this.compensate = compensate;
        }
    }

    /**
     * A saga with typed result collection.
     *
     * Generic parameter {@code T} is the return type of each step's forward action.
     * {@code execute()} returns {@code List<T>}.
     *
     * Usage:
     * <pre>{@code
     * Saga.SagaTyped<String> saga = new Saga.SagaTyped<>();
     * saga.addStep("reserve",
     *     h -> h.durableCall("inv", "Reserve", input).getValue(),
     *     h -> h.durableCall("inv", "Release", input));
     * List<String> results = saga.execute(h);
     * }</pre>
     *
     * @param <T> the result type of each step's forward action
     */
    public static class SagaTyped<T> {
        private final List<StepTyped<T>> steps;

        /** Create a new typed saga with no steps. */
        public SagaTyped() {
            this.steps = new ArrayList<>();
        }

        /**
         * Add a typed step to the saga.
         *
         * @param description human-readable description
         * @param forward     the forward action returning T
         * @param compensate  the compensating action (may be null)
         * @return {@code this} for chaining
         */
        public SagaTyped<T> addStep(String description, java.util.function.Function<HostCalls, T> forward, SagaCompensator compensate) {
            steps.add(new StepTyped<>(description, forward, compensate));
            return this;
        }

        /**
         * Execute all forward steps in order, collecting results.
         *
         * If any step throws an exception, all previously completed steps are
         * compensated in reverse order (skipping null compensations).
         *
         * @param h the HostCalls instance for this workflow execution
         * @return the collected results of all forward steps, in order
         * @throws Exception if a forward step fails and compensation completes
         * @throws RuntimeException if compensation itself fails (the original
         *         forward exception is added as a suppressed exception)
         */
        public List<T> execute(HostCalls h) throws Exception {
            List<StepTyped<T>> completed = new ArrayList<>();
            List<T> results = new ArrayList<>();

            try {
                for (StepTyped<T> step : steps) {
                    T result = step.forward.apply(h);
                    results.add(result);
                    completed.add(step);
                }
            } catch (Exception e) {
                // Compensate completed steps in reverse order.
                Exception compensateFailure = null;
                for (int i = completed.size() - 1; i >= 0; i--) {
                    StepTyped<T> failed = completed.get(i);
                    if (failed.compensate == null) continue;
                    try {
                        failed.compensate.accept(h);
                    } catch (Exception ce) {
                        if (compensateFailure == null) {
                            compensateFailure = ce;
                        } else {
                            compensateFailure.addSuppressed(ce);
                        }
                    }
                }

                if (compensateFailure != null) {
                    RuntimeException rte = new RuntimeException(
                        "saga compensation failed for step '" + completed.get(completed.size() - 1).description
                            + "'; original forward error: " + e.getMessage(), e);
                    rte.addSuppressed(compensateFailure);
                    throw rte;
                }

                throw e;
            }

            return results;
        }
    }

    /**
     * Execute all forward steps in order.  If any step throws an exception,
     * all previously completed steps are compensated in reverse order.
     *
     * @param h the HostCalls instance for this workflow execution
     * @throws Exception if a forward step fails and compensation completes
     * @throws RuntimeException if compensation itself fails (the original
     *         forward exception is added as a suppressed exception)
     */
    public void run(HostCalls h) throws Exception {
        List<Step> completed = new ArrayList<>();

        try {
            for (Step step : steps) {
                step.forward.apply(h);
                completed.add(step);
            }
        } catch (Exception e) {
            // Compensate completed steps in reverse order.
            Exception compensateFailure = null;
            for (int i = completed.size() - 1; i >= 0; i--) {
                try {
                    completed.get(i).compensate.accept(h);
                } catch (Exception ce) {
                    if (compensateFailure == null) {
                        compensateFailure = ce;
                    } else {
                        compensateFailure.addSuppressed(ce);
                    }
                }
            }

            if (compensateFailure != null) {
                RuntimeException rte = new RuntimeException(
                    "saga compensation failed for step '" + completed.get(completed.size() - 1).description
                        + "'; original forward error: " + e.getMessage(), e);
                rte.addSuppressed(compensateFailure);
                throw rte;
            }

            throw e;
        }
    }
}

File: packages/cleat-as/assembly/saga.ts
/**
 * Saga pattern for structured compensation in AssemblyScript.
 *
 * Ported from the Go Saga at durable/runtime.go.
 *
 * ## Usage (with named functions — required by AS constraint #13)
 *
 * ```ts
 * import { HostCalls, Saga } from "@cleat/sdk";
 *
 * function reserveAction(h: HostCalls): string {
 *   let r = h.durableCall("inventory", "Reserve", '{"items":["A","B"]}');
 *   if (r.isError) return r.error;
 *   return ""; // success
 * }
 *
 * function reserveCompensate(h: HostCalls): void {
 *   h.durableCall("inventory", "Release", '{"reservation_id":"..."}');
 * }
 *
 * function myWorkflow(h: HostCalls, input: string): string {
 *   let saga = new Saga();
 *   saga.addStep("reserve", reserveAction, reserveCompensate);
 *   saga.addStep("charge", chargeAction, chargeCompensate);
 *   let result = saga.run(h);
 *   if (result.length > 0) return result; // error
 *   return '{"status":"ok"}';
 * }
 * ```
 *
 * ## Compensation behavior
 *
 * Steps execute in order. If any step fails (returns non-empty error string),
 * all previously completed steps are compensated in reverse order.
 *
 * ## Constraints
 *
 * - `--runtime stub` compatible (no exceptions or try/catch used).
 * - Uses function references, not closures. All action and compensation
 *   functions must be named top-level functions.
 * - Generics are NOT supported for saga result collection. AssemblyScript's
 *   generics system (especially when targeting WASM with `--runtime stub`)
 *   does not support the type-level patterns needed for `Saga<T>` with typed
 *   result collection. Use the Go `SagaTyped[T]` or Java `Saga.SagaTyped<T>`
 *   when typed results are needed.
 *
 * @packageDocumentation
 */

import { HostCalls } from "./host-calls";
import { isWorkflowSuspended } from "./memory";

/**
 * A single saga step with its forward action and compensation function.
 *
 * `forward` returns "" on success or an error message on failure.
 * `compensate` is a cleanup action called when a later step fails.
 * Pass null for compensate if the step has no meaningful compensation
 * (e.g., sending a notification is best-effort).
 */
export class SagaStep {
  constructor(
    /** Human-readable description for logging. */
    public readonly description: string,
    /**
     * Forward action. Takes a HostCalls, returns "" on success
     * or an error message on failure.
     */
    public readonly forward: (h: HostCalls) => string,
    /**
     * Compensation action. Takes a HostCalls. Called when a later
     * step fails. Pass null for steps with no meaningful compensation.
     */
    public readonly compensate: (h: HostCalls) => void | null,
  ) {}
}

/**
 * Provides structured compensation for multi-step operations.
 *
 * Steps execute in order. If any step fails, all previously completed
 * steps are compensated in reverse order.
 *
 * Usage:
 * ```ts
 * let s = new Saga();
 * s.addStep("charge", chargeFn, refundFn);
 * s.addStep("ship", shipFn, cancelShipFn);
 * let err = s.run(h);
 * if (err.length > 0) { /* handle error *\/ }
 * ```
 */
export class Saga {
  private steps: SagaStep[] = [];

  /**
   * Add a step to the saga.
   *
   * @param description - Human-readable description for logging.
   * @param forward     - Main action. Returns "" on success, error message on failure.
   * @param compensate  - Cleanup action. Called on a later step's failure.
   *                      Pass null for best-effort steps with no compensation.
   * @returns The Saga instance for chaining.
   */
  addStep(
    description: string,
    forward: (h: HostCalls) => string,
    compensate: (h: HostCalls) => void | null,
  ): Saga {
    this.steps.push(new SagaStep(description, forward, compensate));
    return this;
  }

  /**
   * Execute all forward steps in order.
   *
   * If any step fails (returns non-empty string), previously completed
   * steps are compensated in reverse order. Null compensate functions
   * are skipped.
   *
   * @param h - HostCalls instance for making workflow calls.
   * @returns "" on success, or an error message on failure.
   *
   * On suspension (durableSleep or awaitChild suspend during execution),
   * the step will set the workflowSuspended flag. The caller should check
   * isWorkflowSuspended() after saga.run() returns and return the SUSPEND_SENTINEL
   * if the workflow suspended mid-execution.
   */
  run(h: HostCalls): string {
    let completed: i32 = 0;

    for (let i: i32 = 0; i < this.steps.length; i++) {
      let step: SagaStep = this.steps[i];
      let result: string = step.forward(h);

      // Check if the workflow suspended during this step
      if (isWorkflowSuspended()) {
        return "";
      }

      // Non-empty result means error
      if (result.length > 0) {
        // Step failed — compensate in reverse order
        this.compensateReverse(h, completed);
        return result;
      }

      completed++;
    }

    return ""; // all steps succeeded
  }

  /**
   * Compensate completed steps in reverse order.
   * Skips null compensate functions.
   */
  private compensateReverse(h: HostCalls, completed: i32): void {
    for (let j: i32 = completed - 1; j >= 0; j--) {
      let cs: SagaStep = this.steps[j];
      if (cs.compensate !== null) {
        cs.compensate(h);
      }
    }
  }
}
